File: src/surface/face/relate/lifecycle.zig
//! relate — the `index` and `status` lifecycle verbs. relate owns its warm
//! state the way `gist` owns the trigram index: a full engine, not a shim.
//!
//!   relate index [--shelf]     build + persist the kinship atlas
//!                              (`kinship.atlas`: one LZJD sketch + one
//!                              structure silhouette per corpus file —
//!                              index/atlas/atlas.zig); `--shelf` also
//!                              rebuilds the codex shelf `quote` reads (the
//!                              same artifact `gist codex build` writes — one
//!                              shelf, two product faces)
//!   relate status [--json]     is the atlas ready, how big, how fresh — and
//!                              is the codex shelf `quote` needs present
//!
//! The anchor is captured BEFORE the corpus read (the T3 convention), so a
//! file touched mid-build reports as changed on the next query's fold.
//! Persistence is atomic (temp-then-rename) — ~10 coworking agents can race
//! `relate index` against a mid-query load and never observe torn bytes.

const std = @import("std");
const corpus_mod = @import("../../../corpus/tree/corpus.zig");
const fresh = @import("../../../corpus/index/trigrams/fresh.zig");
const persist = @import("../../../corpus/index/trigrams/persist.zig");
const atlas_mod = @import("../../../corpus/index/atlas/atlas.zig");
const frag_mod = @import("../../../corpus/index/frag/frag.zig");
const codex_face = @import("../gist/lifecycle/codex.zig");
const cli_args = @import("../../exec/cold/argv/args.zig");
const kinship = @import("kinship.zig");

const nowNs = cli_args.nowNs;
const ms = cli_args.ms;
const Dir = std.Io.Dir;

/// Version of the `status --json` machine contract; bump on breaking change.
pub const schema_version = 1;

/// `relate index [--shelf]` — sketch the index corpus, persist the atlas
/// atomically; `--shelf` also rebuilds the codex shelf from the same read.
pub fn runIndex(gpa: std.mem.Allocator, io: std.Io, argv: []const []const u8) !void {
    const with_shelf = kinship.onlyFlag(argv, "--shelf", "usage: relate index [--shelf]\n");

    const t0 = nowNs(io);
    const built_ns: i64 = @intCast(std.Io.Clock.now(.real, io).nanoseconds);
    const roots = try corpus_mod.resolveRoots(gpa);
    defer corpus_mod.freeRoots(gpa, roots);
    var corpus = try corpus_mod.load(gpa, io, roots);
    defer corpus.deinit();

    const sketches = kinship.buildSketches(gpa, corpus.docs);
    defer gpa.free(sketches);
    const silhouettes = kinship.buildSilhouettes(gpa, corpus.docs);
    defer gpa.free(silhouettes);
    const blob = try atlas_mod.save(gpa, corpus.paths, sketches, silhouettes, built_ns, roots);
    defer gpa.free(blob);
    try persist.writeAtomic(io, atlas_mod.atlasFile(), blob);
    std.debug.print("atlas: {d} files · {d:.1} MiB corpus → {d:.1} MiB atlas · {d:.0} ms → {s}\n", .{
        corpus.docs.len,
        @as(f64, @floatFromInt(corpus.bytes)) / (1 << 20),
        @as(f64, @floatFromInt(blob.len)) / (1 << 20),
        ms(nowNs(io) - t0),
        atlas_mod.atlasFile(),
    });

    // The fragment tier rides the same corpus read + anchor: one silhouette per
    // extracted function, so `relate concepts` answers warm.
    const f0 = nowNs(io);
    var fbuild = try frag_mod.buildAll(gpa, &corpus);
    defer fbuild.deinit();
    const fblob = try frag_mod.save(gpa, &fbuild, built_ns, roots);
    defer gpa.free(fblob);
    try persist.writeAtomic(io, frag_mod.fragFile(), fblob);
    std.debug.print("frag:  {d} fragment(s) → {d:.1} MiB · {d:.0} ms → {s}\n", .{
        fbuild.count(),
        @as(f64, @floatFromInt(fblob.len)) / (1 << 20),
        ms(nowNs(io) - f0),
        frag_mod.fragFile(),
    });

    if (with_shelf) {
        const s0 = nowNs(io);
        const shelf = try codex_face.persistShelf(gpa, io, &corpus, built_ns);
        std.debug.print("shelf: {d:.1} MiB ({d:.2} bits/char) · {d:.0} ms → {s}\n", .{
            @as(f64, @floatFromInt(shelf.bytes)) / (1 << 20),
            shelf.bits_per_char,
            ms(nowNs(io) - s0),
            codex_face.shelfFile(),
        });
    }
}

fn fileBytes(io: std.Io, path: []const u8) ?u64 {
    const st = Dir.cwd().statFile(io, path, .{}) catch return null;
    return @intCast(st.size);
}

const Status = struct {
    schema_version: u8 = schema_version,
    atlas: struct {
        state: enum { ready, unavailable },
        files: usize = 0,
        bytes: u64 = 0,
        stale_files: usize = 0,
        built_unix_ns: i64 = 0,
    },
    frag: struct {
        state: enum { ready, unavailable },
        fragments: usize = 0,
        bytes: u64 = 0,
        stale_files: usize = 0,
        built_unix_ns: i64 = 0,
    },
    shelf: struct {
        state: enum { ready, unavailable },
        bytes: u64 = 0,
    },
};

/// `relate status [--json]` — atlas + shelf readiness. Exit 0 when the atlas
/// is ready (the kinship verbs run warm), 1 when it is missing/corrupt (they
/// still answer, live). Shelf detail beyond presence stays with
/// `gist codex status` — one artifact, one deep report.
pub fn runStatus(gpa: std.mem.Allocator, io: std.Io, argv: []const []const u8) !void {
    const json = kinship.onlyFlag(argv, "--json", "usage: relate status [--json]\n");

    var st = Status{
        .atlas = .{ .state = .unavailable },
        .frag = .{ .state = .unavailable },
        .shelf = .{ .state = .unavailable },
    };
    if (atlas_mod.loadQuiet(gpa, io)) |atl_v| {
        var atl = atl_v;
        defer atl.deinit(gpa);
        st.atlas = .{
            .state = .ready,
            .files = atl.paths.len,
            .bytes = fileBytes(io, atlas_mod.atlasFile()) orelse 0,
            .stale_files = fresh.staleCount(gpa, io, atl.roots, atl.built_ns),
            .built_unix_ns = atl.built_ns,
        };
    }
    if (frag_mod.loadQuiet(gpa, io)) |frag_v| {
        var f = frag_v;
        defer f.deinit(gpa);
        st.frag = .{
            .state = .ready,
            .fragments = f.spans.len,
            .bytes = fileBytes(io, frag_mod.fragFile()) orelse 0,
            .stale_files = fresh.staleCount(gpa, io, f.roots, f.built_ns),
            .built_unix_ns = f.built_ns,
        };
    }
    if (fileBytes(io, codex_face.shelfFile())) |b| st.shelf = .{ .state = .ready, .bytes = b };

    const shelf_line = if (st.shelf.state == .ready) "ready (quote answers)" else "missing — `relate index --shelf` builds it";
    var out: std.ArrayList(u8) = .empty;
    defer out.deinit(gpa);
    if (json) {
        const body = try std.json.Stringify.valueAlloc(gpa, st, .{});
        defer gpa.free(body);
        try out.print(gpa, "{s}\n", .{body});
    } else if (st.atlas.state == .ready) {
        const frag_line = if (st.frag.state == .ready) "ready (concepts answers)" else "missing — `relate index` builds it";
        try out.print(gpa,
            \\relate — kinship atlas {s}
            \\  files sketched  {d}
            \\  atlas           {d:.1} MiB
            \\  changed since   {d} file(s) (folded in at query time)
            \\  fragments       {d} · {d:.1} MiB · {d} changed — {s}
            \\  codex shelf     {s}
            \\
        , .{
            atlas_mod.atlasFile(),
            st.atlas.files,
            @as(f64, @floatFromInt(st.atlas.bytes)) / (1 << 20),
            st.atlas.stale_files,
            st.frag.fragments,
            @as(f64, @floatFromInt(st.frag.bytes)) / (1 << 20),
            st.frag.stale_files,
            frag_line,
            shelf_line,
        });
    } else {
        try out.print(gpa,
            \\relate — no kinship atlas at {s} (verbs answer live; `relate index` builds it)
            \\  codex shelf     {s}
            \\
        , .{ atlas_mod.atlasFile(), shelf_line });
    }
    corpus_mod.emitStdout(out.items);
    std.process.exit(if (st.atlas.state == .ready) 0 else 1);
}

File: bindings/python/gist/index/lifecycle.py
"""Typed lifecycle and capability introspection — what the accelerators know.

Every verb in this package answers correctly with no persisted artifact at all;
warmth is an optimization tier, never a dependency. This module is how a program
*sees* that tier instead of guessing at it: the trigram index behind exact
search (`status`/`index`), and the kinship atlas, fragment index, and codex
shelf behind the compression verbs (`atlas_status`/`atlas_index`).

Two reasons a caller reaches here rather than shelling `gist status`. A long-
running process can decide *once* whether to pay for a build before a batch of
queries, instead of eating a cold walk per call. And a verb that needs an
artifact rather than merely preferring one — `retrieval.quote` and
`compose.provenance` read the shelf — can be preflighted rather than failed.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum
import json
import subprocess
import time
from typing import TYPE_CHECKING, NotRequired, TypedDict

from ..runtime.errors import SearchFailedError


if TYPE_CHECKING:
    import os

    from ..runtime.shell import Output


class IndexState(StrEnum):
    """Availability state in the versioned status schema."""

    READY = "ready"
    UNAVAILABLE = "unavailable"


@dataclass(frozen=True, slots=True)
class IndexStatus:
    """Current persisted-index state reported by ``gist status``."""

    schema_version: int
    state: IndexState
    path: str | None
    paths_file: str | None
    files: int = 0
    trigrams: int = 0
    postings: int = 0
    index_bytes: int = 0
    paths_bytes: int = 0
    anchor_unix_ns: int | None = None
    age_seconds: float | None = None
    roots: tuple[str, ...] = ()
    bound_here: bool = True
    built_over: str | None = None

    @property
    def ready(self) -> bool:
        """Whether a complete persisted pair is available."""
        return self.state is IndexState.READY

    @property
    def freshness_anchor(self) -> bool:
        """Whether changed files can be folded in against a build anchor."""
        return self.anchor_unix_ns is not None and self.bound_here


@dataclass(frozen=True, slots=True)
class FlagCapability:
    """One CLI flag row from the binary's generated schema."""

    spellings: tuple[str, ...]
    note: str | None = None


@dataclass(frozen=True, slots=True)
class Capabilities:
    """Machine-readable feature surface emitted by ``gist --schema``."""

    tool: str
    version: str
    verbs: tuple[str, ...]
    buckets: tuple[tuple[str, tuple[FlagCapability, ...]], ...]
    native_additions: tuple[str, ...]

    def compatibility(self, spelling: str) -> str | None:
        """Compatibility bucket for ``spelling``, or ``None`` when unknown."""
        return next(
            (
                bucket
                for bucket, flags in self.buckets
                if any(spelling in flag.spellings for flag in flags)
            ),
            None,
        )

    def supports(self, spelling: str) -> bool:
        """Whether the binary recognizes ``spelling`` as a search capability."""
        return self.compatibility(spelling) not in (None, "unsupported-fail-loud")


class _IndexPayload(TypedDict):
    path: str
    paths_file: str
    files_indexed: int
    distinct_trigrams: int
    postings: int
    index_bytes: int
    paths_bytes: int


class _FreshnessPayload(TypedDict):
    anchor_unix_ns: int | None
    age_seconds: float | None


class _StatusPayload(TypedDict):
    schema_version: int
    state: str
    index: _IndexPayload | None
    freshness: _FreshnessPayload
    roots: list[str]
    # Added within schema_version 1 (additive); absent from an older binary.
    bound_here: NotRequired[bool]
    built_over: NotRequired[str | None]


class _FlagPayload(TypedDict):
    spellings: list[str]
    note: NotRequired[str]


class _NativePayload(TypedDict):
    native: str


class _CompatibilityPayload(TypedDict):
    buckets: dict[str, list[_FlagPayload]]


class _RipgrepPayload(TypedDict):
    ripgrep_compatibility: _CompatibilityPayload
    native_additions: list[_NativePayload]


class _SchemaPayload(TypedDict):
    tool: str
    version: str
    verbs: dict[str, object]
    search: _RipgrepPayload


def _command(
    argv: list[str],
    *,
    cwd: str | os.PathLike[str] | None,
    timeout: float,
) -> subprocess.CompletedProcess[str]:
    from ..runtime.shell import binary

    try:
        proc = subprocess.run(  # noqa: S603 — fixed executable and argv list
            [binary(), *argv],
            capture_output=True,
            text=True,
            cwd=cwd,
            timeout=timeout,
            check=False,
            stdin=subprocess.DEVNULL,
        )
    except subprocess.TimeoutExpired as exc:
        msg = f"gist {' '.join(argv)} timed out after {timeout}s"
        raise SearchFailedError(msg) from exc
    if proc.returncode != 0:
        raise SearchFailedError(proc.stderr.strip() or f"gist {' '.join(argv)} failed")
    return proc


def parse_status(report: str) -> IndexStatus:
    """Decode the binary's versioned status JSON without scraping prose."""
    payload: _StatusPayload = json.loads(report)
    index = payload["index"]
    freshness = payload["freshness"]
    return IndexStatus(
        schema_version=payload["schema_version"],
        state=IndexState(payload["state"]),
        path=index["path"] if index else None,
        paths_file=index["paths_file"] if index else None,
        files=index["files_indexed"] if index else 0,
        trigrams=index["distinct_trigrams"] if index else 0,
        postings=index["postings"] if index else 0,
        index_bytes=index["index_bytes"] if index else 0,
        paths_bytes=index["paths_bytes"] if index else 0,
        anchor_unix_ns=freshness["anchor_unix_ns"],
        age_seconds=freshness["age_seconds"],
        roots=tuple(payload["roots"]),
        # An artifact directory built over ANOTHER tree reports real counts and
        # a real anchor that describe none of the files here, so every
        # accelerator declines and the answer comes live. Default True: an older
        # binary that never published a binding has no other tree to name.
        bound_here=payload.get("bound_here", True),
        built_over=payload.get("built_over"),
    )


def status(
    *,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = 30.0,
) -> IndexStatus:
    """Inspect the persisted index without mutating it."""
    return parse_status(_command(["status", "--json"], cwd=cwd, timeout=timeout).stdout)


def index(
    *,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = 600.0,
) -> IndexStatus:
    """Build and atomically publish the index, then return its observed state."""
    _command(["index"], cwd=cwd, timeout=timeout)
    result = status(cwd=cwd, timeout=timeout)
    if not result.ready:
        msg = "gist index completed without a readable index"
        raise SearchFailedError(msg)
    return result


def parse_capabilities(payload: str) -> Capabilities:
    """Decode the generated JSON capability manifest."""
    raw: _SchemaPayload = json.loads(payload)
    compat = raw["search"]["ripgrep_compatibility"]["buckets"]
    buckets = tuple(
        (
            name,
            tuple(FlagCapability(tuple(row["spellings"]), row.get("note")) for row in rows),
        )
        for name, rows in compat.items()
    )
    additions = tuple(row["native"] for row in raw["search"]["native_additions"])
    return Capabilities(
        tool=raw["tool"],
        version=raw["version"],
        verbs=tuple(raw["verbs"]),
        buckets=buckets,
        native_additions=additions,
    )


def capabilities(
    *,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = 30.0,
) -> Capabilities:
    """Return the binary's parser-derived capability schema."""
    return parse_capabilities(_command(["--schema"], cwd=cwd, timeout=timeout).stdout)


# ── the compression tier: kinship atlas, fragment index, codex shelf ─────────


@dataclass(frozen=True, slots=True)
class Artifact:
    """One persisted compression artifact. `stale_files` is not damage — a warm answer folds changed files back in from live bytes and stays byte-identical to a cold rebuild, so staleness costs time, never correctness."""

    state: IndexState
    files: int = 0
    fragments: int = 0
    bytes: int = 0
    stale_files: int = 0
    built_unix_ns: int | None = None

    @property
    def ready(self) -> bool:
        """Whether this artifact can accelerate a query at all."""
        return self.state is IndexState.READY

    @property
    def staleness(self) -> float | None:
        """Share of the snapshotted corpus that changed since the anchor, in [0, 1] — the work a warm query will redo live. `None` when this artifact publishes no comparable file population (the fragment index counts fragments, the shelf counts bytes), where `stale_files` is still the raw truth."""
        return self.stale_files / self.files if self.files else None

    @property
    def age_seconds(self) -> float | None:
        """Wall-clock seconds since this artifact was built, or `None` when it was never built."""
        if self.built_unix_ns is None:
            return None
        return max(0.0, time.time() - self.built_unix_ns / 1e9)


@dataclass(frozen=True, slots=True)
class AtlasStatus:
    """What relate's three artifacts can accelerate right now.

    They are independent because the verbs are: `atlas` carries the kinship
    sketches and structure silhouettes (`similar`/`dups`/`clusters`/`echoes` and
    the composed `family`), `fragments` carries the sub-file fragment index
    (`concepts`/`fragments`), and `shelf` carries the corpus codex that
    quotation reads.
    """

    schema_version: int
    atlas: Artifact
    fragments: Artifact
    shelf: Artifact

    @property
    def ready(self) -> bool:
        """Whether kinship queries can run warm — the tier most verbs use."""
        return self.atlas.ready

    @property
    def can_quote(self) -> bool:
        """Whether `retrieval.quote` and `compose.provenance` have their shelf. Unlike every other artifact these two *require* it, so preflight this rather than catching the failure."""
        return self.shelf.ready


def parse_atlas_status(report: str) -> AtlasStatus:
    """Decode relate's versioned status JSON."""
    payload = json.loads(report)
    return AtlasStatus(
        schema_version=int(payload.get("schema_version", 1)),
        atlas=_artifact(payload.get("atlas")),
        fragments=_artifact(payload.get("frag")),
        shelf=_artifact(payload.get("shelf")),
    )


def _artifact(section: object) -> Artifact:
    """Decode one artifact section, defaulting an absent one to `unavailable`."""
    if not isinstance(section, dict):
        return Artifact(IndexState.UNAVAILABLE)
    return Artifact(
        state=IndexState(section.get("state", "unavailable")),
        files=_count(section, "files"),
        fragments=_count(section, "fragments"),
        bytes=_count(section, "bytes"),
        stale_files=_count(section, "stale_files"),
        built_unix_ns=_count(section, "built_unix_ns") or None,
    )


def _count(section: dict[str, object], key: str) -> int:
    """One count field — absent or JSON `null` reads as zero (the shelf publishes no file counts)."""
    value = section.get(key)
    return int(value) if isinstance(value, int | float) else 0


def atlas_status(
    *,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = 60.0,
) -> AtlasStatus:
    """Inspect relate's persisted artifacts without building anything.

    Read-only and cheap. `relate status` exits 1 when the atlas is missing — that
    is a report, not a failure, so this returns an `unavailable` artifact rather
    than raising.
    """
    out = _relate(["status", "--json"], cwd=cwd, timeout=timeout, ok_codes=(0, 1))
    return parse_atlas_status(out.stdout)


def atlas_index(
    *,
    shelf: bool = False,
    cwd: str | os.PathLike[str] | None = None,
    timeout: float = 900.0,
) -> AtlasStatus:
    """Build and publish relate's kinship atlas, then report what is ready.

    Warm kinship costs seconds to build and returns roughly an order of magnitude
    on every later query, so a process about to ask many kinship questions should
    pay this once up front. `shelf=True` additionally builds the codex shelf,
    which is the *only* way `quote`/`provenance` become available — it is much
    larger, so it stays opt-in.

    The corpus is not a parameter: an artifact records the roots it was built
    over, so the build takes them from the tree at `cwd` (or `$GIST_ROOTS`) and
    the *query* verbs narrow within it. Point `$GIST_DIR` elsewhere to keep an
    artifact home of your own.
    """
    _relate(["index", *(["--shelf"] if shelf else [])], cwd=cwd, timeout=timeout)
    return atlas_status(cwd=cwd, timeout=60.0)


def _relate(
    argv: list[str],
    *,
    cwd: str | os.PathLike[str] | None,
    timeout: float,
    ok_codes: tuple[int, ...] = (0,),
) -> Output:
    """Invoke the `relate` binary. Imported at call time because `engine` imports this module for `IndexStatus` — lifecycle types are the lower layer, and the transport is the upper one."""
    from ..runtime.shell import run_verb

    return run_verb("relate", argv, cwd=cwd, timeout=timeout, ok_codes=ok_codes)
